File: web/app/types/playlist.ts
import type { SceneListItem } from './scene';

export interface PlaylistOwner {
    id: number;
    username: string;
}

export interface PlaylistTag {
    id: number;
    name: string;
    color: string;
}

export interface PlaylistThumbnailScene {
    id: number;
    thumbnail_path: string;
}

export interface PlaylistListItem {
    uuid: string;
    name: string;
    description: string | null;
    visibility: string;
    scene_count: number;
    total_duration: number;
    owner: PlaylistOwner;
    tags: PlaylistTag[];
    thumbnail_scenes: PlaylistThumbnailScene[];
    is_liked: boolean;
    like_count: number;
    created_at: string;
    updated_at: string;
}

export interface PlaylistSceneEntry {
    position: number;
    scene: SceneListItem;
    added_at: string;
}

export interface PlaylistResume {
    scene_id: number | null;
    position_s: number;
}

export interface PlaylistDetail extends PlaylistListItem {
    scenes: PlaylistSceneEntry[];
    resume: PlaylistResume | null;
}

export interface PlaylistListResponse {
    data: PlaylistListItem[];
    pagination: {
        page: number;
        limit: number;
        total_items: number;
        total_pages: number;
    };
}

export interface CreatePlaylistInput {
    name: string;
    description?: string;
    visibility?: string;
    tag_ids?: number[];
    scene_ids?: number[];
}

export interface UpdatePlaylistInput {
    name?: string;
    description?: string;
    visibility?: string;
}

export type PlaylistSortOption =
    | 'created_at_desc'
    | 'created_at_asc'
    | 'name_asc'
    | 'name_desc'
    | 'scene_count_desc'
    | 'updated_at_desc';

export type PlaylistOwnerFilter = 'me' | 'all';
export type PlaylistVisibilityFilter = '' | 'public' | 'private' | 'unlisted';

export const PLAYLIST_SORT_OPTIONS: { value: PlaylistSortOption; label: string }[] = [
    { value: 'created_at_desc', label: 'Newest First' },
    { value: 'created_at_asc', label: 'Oldest First' },
    { value: 'name_asc', label: 'Name A-Z' },
    { value: 'name_desc', label: 'Name Z-A' },
    { value: 'scene_count_desc', label: 'Most Scenes' },
    { value: 'updated_at_desc', label: 'Recently Updated' },
];

File: internal/api/v1/response/playlist.go
package response

import (
	"time"

	"goonhub/internal/core"
)

// PlaylistOwnerResponse represents the owner of a playlist
type PlaylistOwnerResponse struct {
	ID       uint   `json:"id"`
	Username string `json:"username"`
}

// PlaylistTagResponse represents a tag on a playlist
type PlaylistTagResponse struct {
	ID    uint   `json:"id"`
	Name  string `json:"name"`
	Color string `json:"color"`
}

// PlaylistThumbnailSceneResponse represents a scene thumbnail in the playlist grid
type PlaylistThumbnailSceneResponse struct {
	ID            uint   `json:"id"`
	ThumbnailPath string `json:"thumbnail_path"`
}

// PlaylistListItemResponse is the API response for a playlist list item
type PlaylistListItemResponse struct {
	UUID            string                           `json:"uuid"`
	Name            string                           `json:"name"`
	Description     *string                          `json:"description"`
	Visibility      string                           `json:"visibility"`
	SceneCount      int64                            `json:"scene_count"`
	TotalDuration   int64                            `json:"total_duration"`
	Owner           PlaylistOwnerResponse            `json:"owner"`
	Tags            []PlaylistTagResponse            `json:"tags"`
	ThumbnailScenes []PlaylistThumbnailSceneResponse `json:"thumbnail_scenes"`
	IsLiked         bool                             `json:"is_liked"`
	LikeCount       int64                            `json:"like_count"`
	CreatedAt       time.Time                        `json:"created_at"`
	UpdatedAt       time.Time                        `json:"updated_at"`
}

// PlaylistSceneEntryResponse is the API response for a scene in a playlist
type PlaylistSceneEntryResponse struct {
	Position int           `json:"position"`
	Scene    SceneListItem `json:"scene"`
	AddedAt  time.Time     `json:"added_at"`
}

// PlaylistResumeResponse is the API response for playlist resume info
type PlaylistResumeResponse struct {
	SceneID   *uint   `json:"scene_id"`
	PositionS float64 `json:"position_s"`
}

// PlaylistDetailResponse is the API response for a full playlist detail
type PlaylistDetailResponse struct {
	PlaylistListItemResponse
	Scenes []PlaylistSceneEntryResponse `json:"scenes"`
	Resume *PlaylistResumeResponse      `json:"resume"`
}

// NewPlaylistListItemResponse converts a service PlaylistListItem to an API response
func NewPlaylistListItemResponse(item core.PlaylistListItem) PlaylistListItemResponse {
	tags := make([]PlaylistTagResponse, len(item.Tags))
	for i, t := range item.Tags {
		tags[i] = PlaylistTagResponse{
			ID:    t.ID,
			Name:  t.Name,
			Color: t.Color,
		}
	}

	thumbnails := make([]PlaylistThumbnailSceneResponse, len(item.ThumbnailScenes))
	for i, ts := range item.ThumbnailScenes {
		thumbnails[i] = PlaylistThumbnailSceneResponse{
			ID:            ts.ID,
			ThumbnailPath: ts.ThumbnailPath,
		}
	}

	return PlaylistListItemResponse{
		UUID:          item.UUID,
		Name:          item.Name,
		Description:   item.Description,
		Visibility:    item.Visibility,
		SceneCount:    item.SceneCount,
		TotalDuration: item.TotalDuration,
		Owner: PlaylistOwnerResponse{
			ID:       item.Owner.ID,
			Username: item.Owner.Username,
		},
		Tags:            tags,
		ThumbnailScenes: thumbnails,
		IsLiked:         item.IsLiked,
		LikeCount:       item.LikeCount,
		CreatedAt:       item.CreatedAt,
		UpdatedAt:       item.UpdatedAt,
	}
}

// NewPlaylistListResponse converts a slice of service PlaylistListItems to API responses
func NewPlaylistListResponse(items []core.PlaylistListItem) []PlaylistListItemResponse {
	result := make([]PlaylistListItemResponse, len(items))
	for i, item := range items {
		result[i] = NewPlaylistListItemResponse(item)
	}
	return result
}

// NewPlaylistDetailResponse converts a service PlaylistDetail to an API response
func NewPlaylistDetailResponse(detail *core.PlaylistDetail) PlaylistDetailResponse {
	scenes := make([]PlaylistSceneEntryResponse, len(detail.Scenes))
	for i, entry := range detail.Scenes {
		scenes[i] = PlaylistSceneEntryResponse{
			Position: entry.Position,
			Scene:    ToSceneListItem(entry.Scene),
			AddedAt:  entry.AddedAt,
		}
	}

	var resume *PlaylistResumeResponse
	if detail.Resume != nil {
		resume = &PlaylistResumeResponse{
			SceneID:   detail.Resume.SceneID,
			PositionS: detail.Resume.PositionS,
		}
	}

	return PlaylistDetailResponse{
		PlaylistListItemResponse: NewPlaylistListItemResponse(detail.PlaylistListItem),
		Scenes:                   scenes,
		Resume:                   resume,
	}
}
